#' Write code for loading libraries
#'
#' @param lib vector of libraries
#'
#' @rdname wrLib
#' @export wrLib
#'
wrLib <- function(lib) {
  oup = ""
  for(iLib in lib){
    oup = paste0(oup, "library(", iLib, ") \n")
  }
  glue::glue(paste0(oup, "\n"))
}

#' Write code for loading objects for server.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrSVload
#' @export wrSVload
#'
wrSVload <- function(prefix) {
  glue::glue('{prefix}conf = readRDS("{prefix}conf.rds")\n',
             '{prefix}def  = readRDS("{prefix}def.rds")\n',
             '{prefix}gene = readRDS("{prefix}gene.rds")\n',
             '{prefix}meta = readRDS("{prefix}meta.rds")\n',
             '\n\n\n\n')
}

#' Write code for fixed portion of server.R
#'
#' @rdname wrSVfix
#' @export wrSVfix
#'
wrSVfix <- function() {
  glue::glue('### Useful stuff \n',
             '# Colour palette \n',
             'cList = list(c("grey85","#FFF7EC","#FEE8C8","#FDD49E","#FDBB84", \n',
             '               "#FC8D59","#EF6548","#D7301F","#B30000","#7F0000"), \n',
             '             c("#4575B4","#74ADD1","#ABD9E9","#E0F3F8","#FFFFBF", \n',
             '               "#FEE090","#FDAE61","#F46D43","#D73027")[c(1,1:9,9)], \n',
             '             c("#FDE725","#AADC32","#5DC863","#27AD81","#21908C", \n',
             '               "#2C728E","#3B528B","#472D7B","#440154")) \n',
             'names(cList) = c("White-Red", "Blue-Yellow-Red", "Yellow-Green-Purple") \n',
             ' \n',
             '# Panel sizes \n',
             'pList = c("400px", "600px", "800px") \n',
             'names(pList) = c("Small", "Medium", "Large") \n',
             'pList2 = c("600px", "800px", "1000px") \n',
             'names(pList2) = c("Small", "Medium", "Large") \n',
             'sList = c(18,24,30) \n',
             'names(sList) = c("Small", "Medium", "Large") \n',
             ' \n',
             '# Function to extract legend \n',
             'g_legend <- function(a.gplot){{  \n',
             '  tmp <- ggplot_gtable(ggplot_build(a.gplot))  \n',
             '  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")  \n',
             '  legend <- tmp$grobs[[leg]]  \n',
             '  legend \n',
             '}}  \n',
             ' \n',
             '# Plot theme \n',
             'sctheme <- function(base_size = 24, XYval = TRUE, Xang = 0, XjusH = 0.5){{ \n',
             '  oupTheme = theme( \n',
             '    text =             element_text(size = base_size, family = "Helvetica"), \n',
             '    panel.background = element_rect(fill = "white", colour = NA), \n',
             '    axis.line =   element_line(colour = "black"), \n',
             '    axis.ticks =  element_line(colour = "black", size = base_size / 20), \n',
             '    axis.title =  element_text(face = "bold"), \n',
             '    axis.text =   element_text(size = base_size), \n',
             '    axis.text.x = element_text(angle = Xang, hjust = XjusH), \n',
             '    legend.position = "bottom", \n',
             '    legend.key =      element_rect(colour = NA, fill = NA) \n',
             '  ) \n',
             '  if(!XYval){{ \n',
             '    oupTheme = oupTheme + theme( \n',
             '      axis.text.x = element_blank(), axis.ticks.x = element_blank(), \n',
             '      axis.text.y = element_blank(), axis.ticks.y = element_blank()) \n',
             '  }} \n',
             '  return(oupTheme) \n',
             '}} \n',
             ' \n',
             '### Common plotting functions \n',
             '# Plot cell information on dimred \n',
             'scDRcell <- function(inpConf, inpMeta, inpdrX, inpdrY, inp1,  \n',
             '                     inpsiz, inpcol, inpord, inpfsz, inpasp, inptxt){{ \n',
             '  # Prepare ggData \n',
             '  ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '                       inpConf[UI == inp1]$ID, inpConf[grp == TRUE]$ID),  \n',
             '                   with = FALSE] \n',
             '  colnames(ggData)[1:3] = c("X", "Y", "val") \n',
             '  if(inpord == "Max-1st"){{ \n',
             '    ggData = ggData[order(val)] \n',
             '  }} else if(inpord == "Min-1st"){{ \n',
             '    ggData = ggData[order(-val)] \n',
             '  }} \n',
             ' \n',
             '  # Do factoring if required \n',
             '  if(!is.na(inpConf[UI == inp1]$fCL)){{ \n',
             '    ggCol = strsplit(inpConf[UI == inp1]$fCL, "\\\\|")[[1]] \n',
             '    if(is.na(inpConf[UI == inp1]$fUI)){{ \n',
             '      ggDes = strsplit(inpConf[UI == inp1]$fID, "\\\\|")[[1]] \n',
             '    }} else {{ \n',
             '      ggDes = strsplit(inpConf[UI == inp1]$fUI, "\\\\|")[[1]] \n',
             '    }} \n',
             '    names(ggCol) = levels(ggData$val); names(ggDes) = levels(ggData$val) \n',
             '    ggLvl = levels(ggData$val)[levels(ggData$val) %in% unique(ggData$val)] \n',
             '    ggData$val = factor(ggData$val, levels = ggLvl) \n',
             '    ggCol = ggCol[ggLvl]; ggDes = ggDes[ggLvl] \n',
             '  }} \n',
             ' \n',
             '  # Actual ggplot \n',
             '  ggOut = ggplot(ggData, aes(X, Y, color = val)) + \n',
             '    geom_point(size = inpsiz, shape = 16) + xlab(inpdrX) + ylab(inpdrY) + \n',
             '    sctheme(base_size = sList[inpfsz], XYval = inptxt) \n',
             '  if(is.na(inpConf[UI == inp1]$fCL)){{ \n',
             '    ggOut = ggOut + scale_color_gradientn("", colours = cList[[inpcol]]) + \n',
             '      guides(color = guide_colorbar(barwidth = 15)) \n',
             '  }} else {{ \n',
             '    ggOut = ggOut + scale_color_manual("", values = ggCol, labels = ggDes) + \n',
             '      guides(color = guide_legend(override.aes = list(size = 5),  \n',
             '                                  nrow = inpConf[UI == inp1]$fRow)) \n',
             '  }} \n',
             '  if(inpasp == "Square") {{ \n',
             '    rat = (max(ggData$X) - min(ggData$X)) / (max(ggData$Y) - min(ggData$Y)) \n',
             '    ggOut = ggOut + coord_fixed(ratio = rat) \n',
             '  }} else if(inpasp == "Fixed") {{ \n',
             '    ggOut = ggOut + coord_fixed() \n',
             '  }} \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             '# Plot gene expression on dimred \n',
             'scDRgene <- function(inpConf, inpMeta, inpdrX, inpdrY, inp1, inpH5, inpGene,   \n',
             '                     inpsiz, inpcol, inpord, inpfsz, inpasp, inptxt){{ \n',
             '  # Prepare ggData \n',
             '  ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '                       inpConf[grp == TRUE]$ID),  \n',
             '                   with = FALSE] \n',
             '  colnames(ggData)[1:2] = c("X", "Y") \n',
             '   \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  ggData$val = h5data$read(args = list(inpGene[inp1], quote(expr=))) \n',
             '  ggData[val < 0]$val = 0 \n',
             '  h5file$close_all() \n',
             '  if(inpord == "Max-1st"){{ \n',
             '    ggData = ggData[order(val)] \n',
             '  }} else if(inpord == "Min-1st"){{ \n',
             '    ggData = ggData[order(-val)] \n',
             '  }} \n',
             '   \n',
             '  # Actual ggplot \n',
             '  ggOut = ggplot(ggData, aes(X, Y, color = val)) + \n',
             '    geom_point(size = inpsiz, shape = 16) + xlab(inpdrX) + ylab(inpdrY) + \n',
             '    sctheme(base_size = sList[inpfsz], XYval = inptxt) +  \n',
             '    scale_color_gradientn(inp1, colours = cList[[inpcol]]) + \n',
             '      guides(color = guide_colorbar(barwidth = 15)) \n',
             '  if(inpasp == "Square") {{ \n',
             '    rat = (max(ggData$X) - min(ggData$X)) / (max(ggData$Y) - min(ggData$Y)) \n',
             '    ggOut = ggOut + coord_fixed(ratio = rat) \n',
             '  }} else if(inpasp == "Fixed") {{ \n',
             '    ggOut = ggOut + coord_fixed() \n',
             '  }} \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             '# Get gene list \n',
             'scGeneList <- function(inp, inpGene){{ \n',
             '  geneList = data.table(gene = unique(trimws(strsplit(inp, ",|;|\n")[[1]])), \n',
             '                        present = TRUE) \n',
             '  geneList[!gene %in% names(inpGene)]$present = FALSE \n',
             '  return(geneList) \n',
             '}} \n',
             ' \n',
             '# Plot gene expression bubbleplot / heatmap \n',
             'scBubbHeat <- function(inpConf, inpMeta, inp, inpGrp, inpPlt,  \n',
             '                       inpH5, inpGene, inpScl, inpRow, inpCol,  \n',
             '                       inpcols, inpfsz, save = FALSE){{ \n',
             '  # Identify genes that are in our dataset \n',
             '  geneList = scGeneList(inp, inpGene) \n',
             '  geneList = geneList[present == TRUE] \n',
             '  validate(need(nrow(geneList) <= 50, "More than 50 genes to plot! Please reduce the gene list!")) \n',
             '  validate(need(nrow(geneList) > 1, "Please input at least 2 genes to plot!")) \n',
             '   \n',
             '  # Prepare ggData \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  ggData = data.table() \n',
             '  for(iGene in geneList$gene){{ \n',
             '    tmp = inpMeta[, c("sampleID", inpConf[grp == TRUE]$ID), with = FALSE] \n',
             '    tmp$grpBy = inpMeta[[inpConf[UI == inpGrp]$ID]] \n',
             '    tmp$geneName = iGene \n',
             '    tmp$val = h5data$read(args = list(inpGene[iGene], quote(expr=))) \n',
             '    ggData = rbindlist(list(ggData, tmp)) \n',
             '  }} \n',
             '  h5file$close_all() \n',
             '   \n',
             '  # Aggregate \n',
             '  ggData$val = expm1(ggData$val) \n',
             '  ggData = ggData[, .(val = mean(val), prop = sum(val>0) / length(sampleID)), \n',
             '                  by = c("geneName", "grpBy")] \n',
             '  ggData$val = log1p(ggData$val) \n',
             '   \n',
             '  # Scale if required \n',
             '  colRange = range(ggData$val) \n',
             '  if(inpScl){{ \n',
             '    ggData[, val:= scale(val), keyby = "geneName"] \n',
             '    colRange = c(-max(abs(range(ggData$val))), max(abs(range(ggData$val)))) \n',
             '  }} \n',
             '   \n',
             '  # hclust row/col if necessary \n',
             '  ggMat = dcast.data.table(ggData, geneName~grpBy, value.var = "val") \n',
             '  tmp = ggMat$geneName \n',
             '  ggMat = as.matrix(ggMat[, -1]) \n',
             '  rownames(ggMat) = tmp \n',
             '  if(inpRow){{ \n',
             '    hcRow = dendro_data(as.dendrogram(hclust(dist(ggMat)))) \n',
             '    ggRow = ggplot() + coord_flip() + \n',
             '      geom_segment(data = hcRow$segments, aes(x=x,y=y,xend=xend,yend=yend)) + \n',
             '      scale_y_continuous(breaks = rep(0, uniqueN(ggData$grpBy)), \n',
             '                         labels = unique(ggData$grpBy), expand = c(0, 0)) + \n',
             '      scale_x_continuous(breaks = seq_along(hcRow$labels$label), \n',
             '                         labels = hcRow$labels$label, expand = c(0, 0.5)) + \n',
             '      sctheme(base_size = sList[inpfsz]) + \n',
             '      theme(axis.title = element_blank(), axis.line = element_blank(), \n',
             '            axis.ticks = element_blank(), axis.text.y = element_blank(), \n',
             '            axis.text.x = element_text(color="white", angle = 45, hjust = 1)) \n',
             '    ggData$geneName = factor(ggData$geneName, levels = hcRow$labels$label) \n',
             '  }} else {{ \n',
             '    ggData$geneName = factor(ggData$geneName, levels = rev(geneList$gene)) \n',
             '  }} \n',
             '  if(inpCol){{ \n',
             '    hcCol = dendro_data(as.dendrogram(hclust(dist(t(ggMat))))) \n',
             '    ggCol = ggplot() + \n',
             '      geom_segment(data = hcCol$segments, aes(x=x,y=y,xend=xend,yend=yend)) + \n',
             '      scale_x_continuous(breaks = seq_along(hcCol$labels$label), \n',
             '                         labels = hcCol$labels$label, expand = c(0.05, 0)) + \n',
             '      scale_y_continuous(breaks = rep(0, uniqueN(ggData$geneName)), \n',
             '                         labels = unique(ggData$geneName), expand=c(0,0)) + \n',
             '      sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) + \n',
             '      theme(axis.title = element_blank(), axis.line = element_blank(), \n',
             '            axis.ticks = element_blank(), axis.text.x = element_blank(), \n',
             '            axis.text.y = element_text(color = "white")) \n',
             '    ggData$grpBy = factor(ggData$grpBy, levels = hcCol$labels$label) \n',
             '  }} \n',
             '   \n',
             '  # Actual plot according to plottype \n',
             '  if(inpPlt == "Bubbleplot"){{ \n',
             '    # Bubbleplot \n',
             '    ggOut = ggplot(ggData, aes(grpBy, geneName, color = val, size = prop)) + \n',
             '      geom_point() +  \n',
             '      sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) +  \n',
             '      scale_x_discrete(expand = c(0.05, 0)) +  \n',
             '      scale_y_discrete(expand = c(0, 0.5)) + \n',
             '      scale_size_continuous("proportion", range = c(0, 8), \n',
             '                            limits = c(0, 1), breaks = c(0.00,0.25,0.50,0.75,1.00)) + \n',
             '      scale_color_gradientn("expression", limits = colRange, colours = cList[[inpcols]]) + \n',
             '      guides(color = guide_colorbar(barwidth = 15)) + \n',
             '      theme(axis.title = element_blank(), legend.box = "vertical") \n',
             '  }} else {{ \n',
             '    # Heatmap \n',
             '    ggOut = ggplot(ggData, aes(grpBy, geneName, fill = val)) + \n',
             '      geom_tile() +  \n',
             '      sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) + \n',
             '      scale_x_discrete(expand = c(0.05, 0)) +  \n',
             '      scale_y_discrete(expand = c(0, 0.5)) + \n',
             '      scale_fill_gradientn("expression", limits = colRange, colours = cList[[inpcols]]) + \n',
             '      guides(fill = guide_colorbar(barwidth = 15)) + \n',
             '      theme(axis.title = element_blank()) \n',
             '  }} \n',
             '     \n',
             '  # Final tidy \n',
             '  ggLeg = g_legend(ggOut) \n',
             '  ggOut = ggOut + theme(legend.position = "none") \n',
             '  if(!save){{ \n',
             '    if(inpRow & inpCol){{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, ggCol, ggRow, widths = c(7,1), heights = c(1,7,2),  \n',
             '                   layout_matrix = rbind(c(3,NA),c(1,4),c(2,NA)))  \n',
             '    }} else if(inpRow){{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, ggRow, widths = c(7,1), heights = c(7,2),  \n',
             '                   layout_matrix = rbind(c(1,3),c(2,NA)))  \n',
             '    }} else if(inpCol){{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, ggCol, heights = c(1,7,2),  \n',
             '                   layout_matrix = rbind(c(3),c(1),c(2)))  \n',
             '    }} else {{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, heights = c(7,2),  \n',
             '                   layout_matrix = rbind(c(1),c(2)))  \n',
             '    }}  \n',
             '  }} else {{ \n',
             '    if(inpRow & inpCol){{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, ggCol, ggRow, widths = c(7,1), heights = c(1,7,2),  \n',
             '                  layout_matrix = rbind(c(3,NA),c(1,4),c(2,NA)))  \n',
             '    }} else if(inpRow){{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, ggRow, widths = c(7,1), heights = c(7,2),  \n',
             '                  layout_matrix = rbind(c(1,3),c(2,NA)))  \n',
             '    }} else if(inpCol){{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, ggCol, heights = c(1,7,2),  \n',
             '                  layout_matrix = rbind(c(3),c(1),c(2)))  \n',
             '    }} else {{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, heights = c(7,2),  \n',
             '                  layout_matrix = rbind(c(1),c(2)))  \n',
             '    }}  \n',
             '  }} \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n',
             '### Start server code \n',
             'shinyServer(function(input, output, session) {{ \n',
             '  ### For all tags and Server-side selectize \n',
             '  observe_helpers() \n',
             ' \n')
}

#' Write code for main block of server.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrSVmain
#' @export wrSVmain
#'
wrSVmain <- function(prefix) {
  glue::glue('   updateSelectizeInput(session, "{prefix}a1inp2", choices = sort(names({prefix}gene)), \n',
             '                       selected = {prefix}def$gene1, server = TRUE) \n',
             '  updateSelectizeInput(session, "{prefix}a3inp1", choices = sort(names({prefix}gene)), \n',
             '                       selected = {prefix}def$gene1, server = TRUE) \n',
             '  updateSelectizeInput(session, "{prefix}a3inp2", choices = sort(names({prefix}gene)), \n',
             '                       selected = {prefix}def$gene2, server = TRUE) \n',
             ' \n',
             ' \n',
             '  ### Plots for tab a1 \n',
             '  output${prefix}a1oup1 <- renderPlot({{ \n',
             '    scDRcell({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp1,  \n',
             '             input${prefix}a1siz1, input${prefix}a1col1, input${prefix}a1ord1, \n',
             '             input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) \n',
             '  }}) \n',
             '  output${prefix}a1oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a1oup1", height = pList[input${prefix}a1psz]) \n',
             '  }}) \n',
             '  output${prefix}a1oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp1,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 6, width = 6, useDingbats = FALSE, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp1,   \n',
             '                      input${prefix}a1siz1, input${prefix}a1col1, input${prefix}a1ord1,  \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) ) \n',
             '  }}) \n',
             '  output${prefix}a1oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp1,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 6, width = 6, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp1,   \n',
             '                      input${prefix}a1siz1, input${prefix}a1col1, input${prefix}a1ord1,  \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) ) \n',
             '  }}) \n',
             '   \n',
             '  output${prefix}a1oup2 <- renderPlot({{ \n',
             '    scDRgene({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp2,  \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}a1siz2, input${prefix}a1col2, input${prefix}a1ord2, \n',
             '             input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) \n',
             '  }}) \n',
             '  output${prefix}a1oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a1oup2", height = pList[input${prefix}a1psz]) \n',
             '  }}) \n',
             '  output${prefix}a1oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 6, width = 6, useDingbats = FALSE, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp2,  \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a1siz2, input${prefix}a1col2, input${prefix}a1ord2, \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) ) \n',
             '  }}) \n',
             '  output${prefix}a1oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 6, width = 6, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp2,  \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a1siz2, input${prefix}a1col2, input${prefix}a1ord2, \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) ) \n',
             '  }}) \n',
             '   \n',
             '   \n',
             '  ### Plots for tab a2 \n',
             '  output${prefix}a2oup1 <- renderPlot({{ \n',
             '    scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp1,  \n',
             '             input${prefix}a2siz1, input${prefix}a2col1, input${prefix}a2ord1, \n',
             '             input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt) \n',
             '  }}) \n',
             '  output${prefix}a2oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a2oup1", height = pList[input${prefix}a2psz]) \n',
             '  }}) \n',
             '  output${prefix}a2oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp1,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 6, width = 6, useDingbats = FALSE, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp1,   \n',
             '                      input${prefix}a2siz1, input${prefix}a2col1, input${prefix}a2ord1,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt) ) \n',
             '  }}) \n',
             '  output${prefix}a2oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp1,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 6, width = 6, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp1,   \n',
             '                      input${prefix}a2siz1, input${prefix}a2col1, input${prefix}a2ord1,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt) ) \n',
             '  }}) \n',
             '   \n',
             '  output${prefix}a2oup2 <- renderPlot({{ \n',
             '    scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp2,  \n',
             '             input${prefix}a2siz2, input${prefix}a2col2, input${prefix}a2ord2, \n',
             '             input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt) \n',
             '  }}) \n',
             '  output${prefix}a2oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a2oup2", height = pList[input${prefix}a2psz]) \n',
             '  }}) \n',
             '  output${prefix}a2oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 6, width = 6, useDingbats = FALSE, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp2,   \n',
             '                      input${prefix}a2siz2, input${prefix}a2col2, input${prefix}a2ord2,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt) ) \n',
             '  }}) \n',
             '  output${prefix}a2oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 6, width = 6, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp2,   \n',
             '                      input${prefix}a2siz2, input${prefix}a2col2, input${prefix}a2ord2,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt) ) \n',
             '  }}) \n',
             '   \n',
             '   \n',
             '  ### Plots for tab a3 \n',
             '  output${prefix}a3oup1 <- renderPlot({{ \n',
             '    scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp1,  \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}a3siz1, input${prefix}a3col1, input${prefix}a3ord1, \n',
             '             input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) \n',
             '  }}) \n',
             '  output${prefix}a3oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a3oup1", height = pList[input${prefix}a3psz]) \n',
             '  }}) \n',
             '  output${prefix}a3oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp1,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 6, width = 6, useDingbats = FALSE, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp1,  \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz1, input${prefix}a3col1, input${prefix}a3ord1, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '  output${prefix}a3oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp1,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 6, width = 6, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp1,  \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz1, input${prefix}a3col1, input${prefix}a3ord1, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '   \n',
             '  output${prefix}a3oup2 <- renderPlot({{ \n',
             '    scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp2,  \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}a3siz2, input${prefix}a3col2, input${prefix}a3ord2, \n',
             '             input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) \n',
             '  }}) \n',
             '  output${prefix}a3oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a3oup2", height = pList[input${prefix}a3psz]) \n',
             '  }}) \n',
             '  output${prefix}a3oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 6, width = 6, useDingbats = FALSE, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp2,  \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz2, input${prefix}a3col2, input${prefix}a3ord2, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '  output${prefix}a3oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 6, width = 6, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp2,  \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz2, input${prefix}a3col2, input${prefix}a3ord2, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '     \n',
             '   \n',
             '  ### Plots for tab b1 \n',
             '  output${prefix}b1oupTxt <- renderUI({{ \n',
             '    geneList = scGeneList(input${prefix}b1inp, {prefix}gene) \n',
             '    if(nrow(geneList) > 50){{ \n',
             '      HTML("More than 50 input genes! Please reduce the gene list!") \n',
             '    }} else {{ \n',
             '      oup = paste0(nrow(geneList[present == TRUE]), " genes OK and will be plotted") \n',
             '      if(nrow(geneList[present == FALSE]) > 0){{ \n',
             '        oup = paste0(oup, "<br/>", \n',
             '                     nrow(geneList[present == FALSE]), " genes not found (", \n',
             '                     paste0(geneList[present == FALSE]$gene, collapse = ", "), ")") \n',
             '      }} \n',
             '      HTML(oup) \n',
             '    }} \n',
             '  }}) \n',
             '  output${prefix}b1oup <- renderPlot({{ \n',
             '    scBubbHeat({prefix}conf, {prefix}meta, input${prefix}b1inp, input${prefix}b1grp, input${prefix}b1plt, \n',
             '               "{prefix}gexpr.h5", {prefix}gene, \n',
             '               input${prefix}b1scl, input${prefix}b1row, input${prefix}b1col, \n',
             '               input${prefix}b1cols, input${prefix}b1fsz) \n',
             '  }}) \n',
             '  output${prefix}b1oup.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}b1oup", height = pList2[input${prefix}b1psz]) \n',
             '  }}) \n',
             '  output${prefix}b1oup.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b1plt,"_",input${prefix}b1grp,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 10, width = 10, \n',
             '      plot = scBubbHeat({prefix}conf, {prefix}meta, input${prefix}b1inp, input${prefix}b1grp, input${prefix}b1plt, \n',
             '                        "{prefix}gexpr.h5", {prefix}gene, \n',
             '                        input${prefix}b1scl, input${prefix}b1row, input${prefix}b1col, \n',
             '                        input${prefix}b1cols, input${prefix}b1fsz, save = TRUE) ) \n',
             '  }}) \n',
             '  output${prefix}b1oup.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b1plt,"_",input${prefix}b1grp,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 10, width = 10, \n',
             '      plot = scBubbHeat({prefix}conf, {prefix}meta, input${prefix}b1inp, input${prefix}b1grp, input${prefix}b1plt, \n',
             '                        "{prefix}gexpr.h5", {prefix}gene, \n',
             '                        input${prefix}b1scl, input${prefix}b1row, input${prefix}b1col, \n',
             '                        input${prefix}b1cols, input${prefix}b1fsz, save = TRUE) ) \n',
             '  }}) \n',
             '   \n',
             '   \n',
             '   \n')
}

#' Write code for final portion of server.R
#'
#' @rdname wrSVend
#' @export wrSVend
#'
wrSVend <- function() {
  glue::glue('   \n',
             '   \n',
             '}}) \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n')
}

#' Write code for loading objects for ui.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrUIload
#' @export wrUIload
#'
wrUIload <- function(prefix) {
  glue::glue('{prefix}conf = readRDS("{prefix}conf.rds")\n',
             '{prefix}def  = readRDS("{prefix}def.rds")\n',
             '\n\n\n\n')
}

#' Write code for front portion of ui.R
#'
#' @param title shiny app title
#'
#' @rdname wrUIsingle
#' @export wrUIsingle
#'
wrUIsingle <- function(title) {
  glue::glue('### Start server code \n',
             'shinyUI(fluidPage( \n',
             '### HTML formatting of error messages \n',
             'tags$head(tags$style(HTML(".shiny-output-error-validation {{color: red; font-weight: bold;}}"))), \n',
             'list(tags$style(HTML(".navbar-default .navbar-nav {{ font-weight: bold; font-size: 16px; }}"))), \n',
             ' \n',
             '   \n',
             '### Page title \n',
             'titlePanel("{title}"),  \n',
             'navbarPage( \n',
             '  NULL,  \n',
             ' \n')
}

#' Write code for main block of ui.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrUImain
#' @export wrUImain
#'
wrUImain <- function(prefix) {
  glue::glue('  ### Tab1.a1: cellInfo vs geneExpr on dimRed \n',
             '  tabPanel( \n',
             '    HTML("Cell info vs Gene expr"), \n',
             '    h4("Cell information vs gene expression on reduced dimensions"), \n',
             '    "In this tab, users can visualise both cell information and gene ",  \n',
             '    "expression side-by-side on low-dimensional representions.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, h4("Dimension Reduction"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a1drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '            selectInput("{prefix}a1drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                        selected = {prefix}def$dimred[2])) \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, actionButton("{prefix}a1tog0", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a1tog0 % 2 == 1", \n',
             '          fluidRow( \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a1psz", "Plot size:", \n',
             '                              choices = c("Small", "Medium", "Large"), \n',
             '                              selected = "Medium", inline = TRUE), \n',
             '              radioButtons("{prefix}a1fsz", "Font size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE) \n',
             '            ), \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a1asp", "Aspect ratio:", \n',
             '                              choices = c("Square", "Fixed", "Free"), \n',
             '                              selected = "Square", inline = TRUE), \n',
             '              checkboxInput("{prefix}a1txt", "Show axis text", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ) \n',
             '      )  # End of column (6 space) \n',
             '    ),   # End of fluidRow (4 space) \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, style="border-right: 2px solid black", h4("Cell information"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a1inp1", "Cell information:", \n',
             '                           choices = {prefix}conf$UI, \n',
             '                           selected = {prefix}conf[default == 1]$UI) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Cell information to colour cells by", \n',
             '                     content = c("Select cell information to colour cells", \n',
             '                                 "- Categorical covariates have a fixed colour palette", \n',
             '                                 paste0("- Continuous covariates are coloured in a ",  \n',
             '                                        "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a1tog1", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a1tog1 % 2 == 1", \n',
             '              sliderInput("{prefix}a1siz1", "Point size:", \n',
             '                          min = 0, max = 3, value = 1, step = 0.25), \n',
             '              radioButtons("{prefix}a1col1", "Colour (Continuous data):", \n',
             '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '                           selected = "Blue-Yellow-Red"), \n',
             '              radioButtons("{prefix}a1ord1", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original"), \n',
             '                           selected = "Original", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a1oup1.ui"))), \n',
             '        downloadButton("{prefix}a1oup1.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a1oup1.png", "Download PNG") \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, h4("Gene expression"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a1inp2", "Gene name:", choices=NULL) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Gene expression to colour cells by", \n',
             '                     content = c("Select gene to colour cells by gene expression", \n',
             '                                 paste0("- Gene expression are coloured in a ", \n',
             '                                        "White-Red colour scheme which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a1tog2", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a1tog2 % 2 == 1", \n',
             '              sliderInput("{prefix}a1siz2", "Point size:", \n',
             '                          min = 0, max = 3, value = 1, step = 0.25), \n',
             '              radioButtons("{prefix}a1col2", "Colour:", \n',
             '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '                           selected = "White-Red"), \n',
             '              radioButtons("{prefix}a1ord2", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original"), \n',
             '                           selected = "Max-1st", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ) , \n',
             '        fluidRow(column(12, uiOutput("{prefix}a1oup2.ui"))), \n',
             '        downloadButton("{prefix}a1oup2.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a1oup2.png", "Download PNG") \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  ),     # End of tab (2 space) \n',
             ' \n',
             '  ### Tab1.a2: cellInfo vs cellInfo on dimRed \n',
             '  tabPanel( \n',
             '    HTML("Cell info vs Cell info"), \n',
             '    h4("Cell information vs cell information on dimension reduction"), \n',
             '    "In this tab, users can visualise two cell informations side-by-side ", \n',
             '    "on low-dimensional representions.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, h4("Dimension Reduction"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a2drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '            selectInput("{prefix}a2drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                        selected = {prefix}def$dimred[2])) \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, actionButton("{prefix}a2tog0", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a2tog0 % 2 == 1", \n',
             '          fluidRow( \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a2psz", "Plot size:", \n',
             '                              choices = c("Small", "Medium", "Large"), \n',
             '                              selected = "Medium", inline = TRUE), \n',
             '              radioButtons("{prefix}a2fsz", "Font size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE) \n',
             '            ), \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a2asp", "Aspect ratio:", \n',
             '                              choices = c("Square", "Fixed", "Free"), \n',
             '                              selected = "Square", inline = TRUE), \n',
             '              checkboxInput("{prefix}a2txt", "Show axis text", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ) \n',
             '      )  # End of column (6 space) \n',
             '    ),   # End of fluidRow (4 space) \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, style="border-right: 2px solid black", h4("Cell information 1"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a2inp1", "Cell information:", \n',
             '                           choices = {prefix}conf$UI, \n',
             '                           selected = {prefix}conf[default == 1]$UI) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Cell information to colour cells by", \n',
             '                     content = c("Select cell information to colour cells", \n',
             '                                 "- Categorical covariates have a fixed colour palette", \n',
             '                                 paste0("- Continuous covariates are coloured in a ",  \n',
             '                                        "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a2tog1", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a2tog1 % 2 == 1", \n',
             '              sliderInput("{prefix}a2siz1", "Point size:", \n',
             '                          min = 0, max = 3, value = 1, step = 0.25), \n',
             '              radioButtons("{prefix}a2col1", "Colour (Continuous data):", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "Blue-Yellow-Red"), \n',
             '              radioButtons("{prefix}a2ord1", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original"), \n',
             '                           selected = "Original", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a2oup1.ui"))), \n',
             '        downloadButton("{prefix}a2oup1.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a2oup1.png", "Download PNG") \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, h4("Cell information 2"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a2inp2", "Cell information:", \n',
             '                           choices = {prefix}conf$UI, \n',
             '                           selected = {prefix}conf[default == 2]$UI) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Cell information to colour cells by", \n',
             '                     content = c("Select cell information to colour cells", \n',
             '                                 "- Categorical covariates have a fixed colour palette", \n',
             '                                 paste0("- Continuous covariates are coloured in a ",  \n',
             '                                        "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a2tog2", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a2tog2 % 2 == 1", \n',
             '              sliderInput("{prefix}a2siz2", "Point size:", \n',
             '                          min = 0, max = 3, value = 1, step = 0.25), \n',
             '              radioButtons("{prefix}a2col2", "Colour (Continuous data):", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "Blue-Yellow-Red"), \n',
             '              radioButtons("{prefix}a2ord2", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original"), \n',
             '                           selected = "Original", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a2oup2.ui"))), \n',
             '        downloadButton("{prefix}a2oup2.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a2oup2.png", "Download PNG") \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  ),     # End of tab (2 space) \n',
             '   \n',
             '  ### Tab1.a3: geneExpr vs geneExpr on dimRed \n',
             '  tabPanel( \n',
             '    HTML("Gene expr vs Gene expr"), \n',
             '    h4("Gene expression vs gene expression on dimension reduction"), \n',
             '    "In this tab, users can visualise two gene expressions side-by-side ", \n',
             '    "on low-dimensional representions.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, h4("Dimension Reduction"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a3drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '            selectInput("{prefix}a3drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                        selected = {prefix}def$dimred[2])) \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, actionButton("{prefix}a3tog0", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a3tog0 % 2 == 1", \n',
             '          fluidRow( \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a3psz", "Plot size:", \n',
             '                              choices = c("Small", "Medium", "Large"), \n',
             '                              selected = "Medium", inline = TRUE), \n',
             '              radioButtons("{prefix}a3fsz", "Font size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE) \n',
             '            ), \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a3asp", "Aspect ratio:", \n',
             '                              choices = c("Square", "Fixed", "Free"), \n',
             '                              selected = "Square", inline = TRUE), \n',
             '              checkboxInput("{prefix}a3txt", "Show axis text", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ) \n',
             '      )  # End of column (6 space) \n',
             '    ),   # End of fluidRow (4 space) \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, style="border-right: 2px solid black", h4("Gene expression 1"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a3inp1", "Gene name:", choices=NULL) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Gene expression to colour cells by", \n',
             '                     content = c("Select gene to colour cells by gene expression", \n',
             '                                 paste0("- Gene expression are coloured in a ", \n',
             '                                        "White-Red colour scheme which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a3tog1", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a3tog1 % 2 == 1", \n',
             '              sliderInput("{prefix}a3siz1", "Point size:", \n',
             '                          min = 0, max = 3, value = 1, step = 0.25), \n',
             '              radioButtons("{prefix}a3col1", "Colour:", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "White-Red"), \n',
             '              radioButtons("{prefix}a3ord1", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original"), \n',
             '                           selected = "Max-1st", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a3oup1.ui"))), \n',
             '        downloadButton("{prefix}a3oup1.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a3oup1.png", "Download PNG") \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, h4("Gene expression 2"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a3inp2", "Gene name:", choices=NULL) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Gene expression to colour cells by", \n',
             '                     content = c("Select gene to colour cells by gene expression", \n',
             '                                 paste0("- Gene expression are coloured in a ", \n',
             '                                        "White-Red colour scheme which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a3tog2", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a3tog2 % 2 == 1", \n',
             '              sliderInput("{prefix}a3siz2", "Point size:", \n',
             '                          min = 0, max = 3, value = 1, step = 0.25), \n',
             '              radioButtons("{prefix}a3col2", "Colour:", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "White-Red"), \n',
             '              radioButtons("{prefix}a3ord2", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original"), \n',
             '                           selected = "Max-1st", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a3oup2.ui"))), \n',
             '        downloadButton("{prefix}a3oup2.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a3oup2.png", "Download PNG") \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  ),     # End of tab (2 space) \n',
             ' \n',
             '  ### Tab1.b1: Multiple gene expr \n',
             '  tabPanel( \n',
             '    HTML("Bubbleplot / Heatmap"), \n',
             '    h4("Gene expression bubbleplot / heatmap"), \n',
             '    "In this tab, users can visualise the gene expression patterns of ", \n',
             '    "multiple genes grouped by categorical cell information (e.g. library / cluster).", br(), \n',
             '    "The normalised expression are averaged, log-transformed and then plotted.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        3, style="border-right: 2px solid black", \n',
             '        textAreaInput("{prefix}b1inp", HTML("List of gene names <br /> \n',
             '                                          (Max 50 genes, separated <br /> \n',
             '                                           by , or ; or newline):"), \n',
             '                      height = "200px", \n',
             '                      value = paste0({prefix}def$genes, collapse = ", ")) %>% \n',
             '          helper(type = "inline", size = "m", fade = TRUE, \n',
             '                 title = "List of genes to plot on bubbleplot / heatmap", \n',
             '                 content = c("Input genes to plot", \n',
             '                             "- Maximum 50 genes (due to ploting space limitations)", \n',
             '                             "- Genes should be separated by comma, semicolon or newline")), \n',
             '        selectInput("{prefix}b1grp", "Group by:", \n',
             '                    choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                    selected = {prefix}conf[grp == TRUE & default > 0]$UI[1]) %>% \n',
             '          helper(type = "inline", size = "m", fade = TRUE, \n',
             '                 title = "Cell information to group cells by", \n',
             '                 content = c("Select categorical cell information to group cells by", \n',
             '                             "- Single cells are grouped by this categorical covariate", \n',
             '                             "- Plotted as the X-axis of the bubbleplot / heatmap")), \n',
             '        radioButtons("{prefix}b1plt", "Plot type:", \n',
             '                     choices = c("Bubbleplot", "Heatmap"), \n',
             '                     selected = "Bubbleplot", inline = TRUE), \n',
             '        checkboxInput("{prefix}b1scl", "Scale gene expression", value = TRUE), \n',
             '        checkboxInput("{prefix}b1row", "Cluster rows (genes)", value = TRUE), \n',
             '        checkboxInput("{prefix}b1col", "Cluster columns (samples)", value = FALSE), \n',
             '        br(), \n',
             '        actionButton("{prefix}b1tog", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}b1tog % 2 == 1", \n',
             '          radioButtons("{prefix}b1cols", "Colour scheme:", \n',
             '                       choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                   "Yellow-Green-Purple"), \n',
             '                       selected = "Blue-Yellow-Red"), \n',
             '          radioButtons("{prefix}b1psz", "Plot size:", \n',
             '                       choices = c("Small", "Medium", "Large"), \n',
             '                       selected = "Medium", inline = TRUE), \n',
             '          radioButtons("{prefix}b1fsz", "Font size:", \n',
             '                       choices = c("Small", "Medium", "Large"), \n',
             '                       selected = "Medium", inline = TRUE)) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        9, h4(htmlOutput("{prefix}b1oupTxt")), \n',
             '        uiOutput("{prefix}b1oup.ui"), \n',
             '        downloadButton("{prefix}b1oup.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}b1oup.png", "Download PNG") \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  )      # End of tab (2 space) \n',
             '   \n')
}

#' Write code for final portion of ui.R
#'
#' @param footnote shiny app footnote
#' 
#' @rdname wrUIend
#' @export wrUIend
#'
wrUIend <- function(footnote) {
  glue::glue('   \n',
             '   \n',
             'br(), \n',
             'p({footnote}), \n',
             'br(),br(),br(),br(),br() \n',
             '))) \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n')
}




